### PR TITLE
Adds the LayoutNG post

### LINKED ISSUE #1506: RenderingNG deep-dive: LayoutNG
Draft document to turn into markdown here:

https://docs.google.com/document/d/1h3NrkNE2ZvrYmx27UOvtfPNx7mKswGvf48XSlLnTilg/edit#

@rachelandrew

### PR DESCRIPTION
**Pull request by rachelandrew**
Fixes #1506 

Latest in the series of RenderingNG posts.

### CODE REVIEW

File: site/en/blog/layoutng/index.md [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+new Unicode bidirectional algorithm using ICU, a lot of correctness fixes, and more.
+
+## Types of layout bugs
+
+Layout bugs broadly speaking fall into four different categories, 
+each with different root causes.
+
+### Correctness
+
+When we think about bugs in the rendering system, we are typically thinking about correctness,
```
Comment (chrishtr) [inline site/en/blog/layoutng/index.md:96]:
```suggestion
When we think about bugs in the rendering system, we typically think about correctness, 
```

File: site/en/blog/layoutng/index.md [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+}
+```
+
+A fix for this type of problem would typically cause a severe performance regression, 
+(see over-invalidation below), and was very delicate to get correct.
+
+Today (as described above) we have an immutable parent constraints object which describes all the inputs from the parent layout to the child. 
+We store this with the resulting immutable fragment. 
+Due to this, 
+we have a centralized place where we compare or "diff" these two inputs to determine if the child needs to have another layout pass performed.
```
Comment (chrishtr) [inline site/en/blog/layoutng/index.md:156]:
```suggestion
we have a centralized place where we *diff* these two inputs to determine if the child needs to have another layout pass performed. 
```

File: site/en/blog/layoutng/index.md [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+  <figcaption>
+    A fixed width/height element doesn't care if the available-size given to it increases, however a percentage-based width/height does. The <em>available-size</em> is represented on the <em>Parent Constraints</em> object, and as part of the diffing algorithm will perform this optimization.
+  </figcaption>
+</figure>
+
+The diffing code for the above example is:
+
+```cpp
+if (width.IsPercent()) {
+  if (old_constraints.WidthPercentageSize() != new_constraints.WidthPercetageSize())
```
Comment (chrishtr) [inline site/en/blog/layoutng/index.md:180]: Line wrap to avoid horizontal scroll.

File: site/en/blog/layoutng/index.md [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+  <figcaption>
+    The video and <a href="https://hysteresis-layout.glitch.me/">demo</a> show a hysteresis bug in Chrome 92 and below. It is fixed in Chrome 93.</figcaption>
+</figure>
+
+With our previous mutable tree, 
+it was incredibly easy to introduce bugs like this. 
+If the code made the mistake of reading the size or position of an object at the incorrect time or stage 
+(as we didn't "clear" the previous size or position for example), 
+we would immediately add a subtle hysteresis bug. 
+These bugs typically don't appear in testing as the majority of tests focus on a single layout and render pass.
```
Comment (chrishtr) [inline site/en/blog/layoutng/index.md:209]:
```suggestion
These bugs typically don't appear in testing as the majority of tests focus on a single layout and render. 
```

File: site/en/blog/layoutng/index.md [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+
+### Over-invalidation and performance
+
+This is the direct opposite of the under-invalidation class of bugs. 
+Often when fixing an under-invalidation bug we'd trigger a performance cliff.
+
+We often had to make difficult choices favouring correctness over performance. 
+In the next section we'll dive deeper into how we mitigated these types of performance issues.
+
+Rise of the two-pass layouts and performance cliffs
```
Comment (chrishtr) [inline site/en/blog/layoutng/index.md:237]:
```suggestion
#### Rise of the two-pass layouts and performance cliffs
```

File: site/en/blog/layoutng/index.md [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+
+#### Rise of the two-pass layouts and performance cliffs
+
+Flex and grid layout represented a shift in the expressiveness of layouts on the web. 
+However, these algorithms were fundamentally different from the block layout algorithm that came before them.
+
+Block layout (in almost all cases) only requires the engine to perform layout on all its children exactly once. 
+This is great for performance, but ends up not being as expressive as web developers want.
+
+As an incredibly simple example,
```
Comment (chrishtr) [inline site/en/blog/layoutng/index.md:245]:
```suggestion
For example, 
```

File: site/en/blog/layoutng/index.md [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+  child->ForceLayout();
+}
+
+A fix for this type of bug would typically be:
+
+if (/* some very complicated statement */ ||
+    /* another very complicated statement */) {
+  child->ForceLayout();
+}
+```
```
Comment (bfgeek) [inline site/en/blog/layoutng/index.md:148]: This block might be better to split up into two, e.g.

```cpp
if (/* some very complicated statement */) {
  child->ForceLayout();
}
```

A fix for this type of bug would typically be:

```cpp
if (/* some very complicated statement */ ||
    /* another very complicated statement */) {
  child->ForceLayout();
}
```

File: site/en/blog/layoutng/index.md [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+
+The diffing code for the above example is:
+
+```cpp
+if (width.IsPercent()) {
+  if (old_constraints.WidthPercentageSize() != new_constraints.WidthPercetageSize())
+   return kNeedsLayout;
+}
+if (height.IsPercent()) {
+  if (old_constraints.HeightPercentageSize() != new_constraints.HeightPercetageSize())
```
Comment (bfgeek) [inline site/en/blog/layoutng/index.md:184]: here and above `HeightPercetageSize` -> `HeightPercentageSize`

(my bad :P )

**Comment by bfgeek** [comment]
lgtm modulo comments.

**Comment by rachelandrew** [comment]
@chrishtr @bfgeek changes made, let me know when you want to publish.

File: site/en/blog/layoutng/index.md [hunk 9]
```diff
@@ -0,0 +1,342 @@
+---
+layout: 'layouts/blog-post.njk'
+title: 'RenderingNG deep-dive: LayoutNG'
+description: >
+  How the large architecture change of LayoutNG reduces and mitigates various types of bugs and performance issues.
+thumbnail: image/kheDArv5csY6rvQUJDbWRscckLr1/PSZ58CfBFUUNasONGnHg.jpg
+alt: The conceptual model of the new layout architecture.
+date: 2021-10-06
```
Comment (rachelandrew) [inline site/en/blog/layoutng/index.md:8]:
```suggestion
date: 2021-10-08
```

**Pull request merged**